tc_build/rust.py

```diff
@@ -0,0 +1,165 @@
+                        "rustfmt",
+                        "src",
+                    ]
+                    optimized-compiler-builtins = true
+
+                    [install]
+                    prefix = "{install_folder}"
+                    sysconfdir = "etc"
+
+                    [rust]
+                    debug = {toml_boolean(self.debug)}
+                    codegen-tests = false
+
+                    [target.x86_64-unknown-linux-gnu]
+                    llvm-config = "{self.llvm_install_folder}/bin/llvm-config"
+                '''))
+
+        self.clean_build_folder()
+
+    def show_install_info(self):
+        # Installation folder is optional, show build folder as the
+        # installation location in that case.
+        install_folder = self.folders.install if self.folders.install else self.folders.build
+        if not install_folder:
+            raise RuntimeError('Installation folder not set?')
+        if not install_folder.exists():
+            raise RuntimeError('Installation folder does not exist, run build()?')
+        if not (bin_folder := Path(install_folder, 'bin')).exists():
+            raise RuntimeError('bin folder does not exist in installation folder, run build()?')
+
+        tc_build.utils.print_header('Rust installation information')
+        install_info = (f"Toolchain is available at: {install_folder}\n\n"
+                        'To use, either run:\n\n'
+                        f"\t$ export PATH={bin_folder}:$PATH\n\n"
+                        'or add:\n\n'
+                        f"\tPATH={bin_folder}:$PATH\n\n"
+                        'before the command you want to use this toolchain.\n')
+        print(install_info)
+
+        for tool in ['rustc', 'rustdoc', 'rustfmt', 'clippy-driver', 'cargo']:
+            if (binary := Path(bin_folder, tool)).exists():
+                subprocess.run([binary, '--version', '--verbose'], check=True)
+                print()
+        tc_build.utils.flush_std_err_out()
+
+
+class RustSourceManager:
+
+    def __init__(self, repo):
+        self.repo = repo
+
+    def download(self, ref):
+        if self.repo.exists():
+            return
+
+        tc_build.utils.print_header('Downloading Rust')
+
+        git_clone = ['git', 'clone', 'https://github.com/rust-lang/rust.git', self.repo]
+
+        subprocess.run(git_clone, check=True)
+
+        self.git(['checkout', ref])
+
+    def git(self, cmd, capture_output=False):
+        return subprocess.run(['git', *cmd],
+                              capture_output=capture_output,
+                              check=True,
+                              cwd=self.repo,
+                              text=True)
+
+    def git_capture(self, cmd):
+        return self.git(cmd, capture_output=True).stdout.strip()
+
+    def ref_exists(self, ref):
+        try:
+            self.git(['show-branch', ref])
+        except subprocess.CalledProcessError:
+            return False
+        return True
```

Comment:
We could introduce `GitSourceManager` in `lkt/source.py` that has these and `is_shallow()` from `lkt/llvm.py` so that both pieces of code could inherit them and deduplicate these. It may be possible to include `download()` and `update()` with a little refactoring but I am fine with throwing `NotImplementedError` in the base class and leaving it up to the subclasses to implement, even with a little more duplication.

Reply:
Resolving as you did it and I am putting the commit here (thanks!).